### ISSUE TITLE
Coinscope bitcoin handshake outdated as of 0.14.0

### ISSUE THREAD

**Issue by srya-b** [jameslitton/coinscope#3]
According to the release notes for Bitcoin 0.14.0 [here](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.14.0.md) nodes now disconnect if they don't receive a `verack` from a peer within 60 seconds of connecting to the peer. The coinscope that I have been using (and as far as I can tell, this repository as well) does not send a `verack` as part of the handshake. Hence, there are ~2500 nodes to which coinscope is unable to connect.

This is the specific release note:
#9715 `b08656e` Disconnect peers which we do not receive VERACKs from within 60 sec (TheBlueMatt)

As a temporary fix I have been using a python script that uses the coinscope control hooks and other python scripts to connect to the connector and logserver. Here is an example of the high level verack script: 

```python
def main():
    connector = ConnectorSocket('/var/coinscope/bitcoin-testnet/bitcoin_control')
    logsock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM, 0)
    logsock.connect('/var/coinscope/bitcoin-testnet/logger/clients/bitcoin_msg')

    for log in read_logs(logsock):
        try:
            msg = MsgSerializable.stream_deserialize(StringIO(log.bitcoin_msg))
        except:
            continue
               
        if msg is None: continue
        if msg.command == 'version' and not log.is_sender:
            # Send verack to anyone that sends us a version message
            # only reason we'd get a version message is if we start
            # a connection with a node and the handshake is initiated
            try:
                dmap = dict((hid,ip) for ip, hid in connector.get_cxn())
                print '[%s] Sending verack to' % unix2str(log.timestamp), dmap[log.handle_id]
            except:
                print '[%s] Sending verack to' % unix2str(log.timestamp), log.handle_id
            rid = connector.register_verack()
            connector.send_to_node(rid, log.handle_id)

if __name__ == '__main__':
    main() 
```
There will be a repository that has other necessary python scripts that handle connecting to the connector and collecting logs from the logserver, and I'll put a link to it here soon.

**Comment by jameslitton** [comment]
This doesn't seem like a protocol change as far as I can tell. It looks like, in fact, the connector does try to send the verack.

https://github.com/jameslitton/coinscope/blob/master/connector/src/bitcoin_handler.cpp#L514

Maybe something else is going on that is affecting this. I'll run it against a local bitcoin node to test how a queued message could be getting dropped or if some other change in the protocol means this part of the handshake code doesn't get hit.

**Comment by sbaks0820** [comment]
Oh, that's strange. I just noticed in the debug logs of newer nodes that the verack never showed up.

**Comment by sbaks0820** [comment]
So the comment right next to that case says that this is the case where they initiate the handshake. Perhaps that's the issue that we aren't doing `verack` when we initiate handshake?

**Comment by jameslitton** [comment]
yeah, very simple to add a queued verack in the other case. I have to pull a current bitcoin client and test against it. I planned to do it this weekend but didn't get around to it.

**Comment by amiller** [comment]
Don't worry about this @jameslitton, we'll take care of trying this in a fork and testing it.

**Comment by jameslitton** [comment]
Okay. If it's okay to do the verack before they sent their verack, you can just inline the queued message into RECV_VERSION_INIT by throwing

		const struct sockaddr_in * external = external_addr();
		struct combined_version vers(get_version(user_agent(), remote_addr, *external));
		unique_ptr<struct packed_message> vmsg(get_message("version", vers.as_buffer(), vers.size));

		append_for_write(move(vmsg));
		append_for_write(get_message("verack"));

into it. No need to change the state flag that gets set. If the specific sequencing of messaging matters (e.g., they send a verack before we respond with one) then it's a teeny bit more work.
